prototypes/coin-selection/CoinSelection.hs

```diff
@@ -0,0 +1,278 @@
+    -- ^ Predicate: If a given 'TxIn' were added into the given 'TxOut', would
+    -- the size be less than the maximum output size.
+    , isTxOutValueOk       :: tx -> Value tx -> Bool
+    -- ^ Predicate: is 'TxOut' sufficiently large (i.e. less than min utxo)?
+    , isOurAddress         :: Address tx -> AddressOwner
+    -- ^ Classifies addresses as either the recipient's or our own.
+    } deriving Generic
+
+-- | How to choose inputs for a transaction.
+data Selector
+    = Sweep
+    -- ^ Use all possible unspent transaction inputs, regardless of the
+    -- requested payment amount. This is called a wallet "sweep" -- where all
+    -- coins from the wallet are shifted together.
+    | RandomImprove
+    -- ^ Use the /Random-Improve-MA/ algorithm for selecting inputs. This will
+    -- select approximately double the amount necessary for the payment.
+    deriving (Show, Read, Eq, Enum, Generic)
+
+-- | How to assign excess value to change outputs.
+data Change
+    = Balanced
+    -- ^ Value is distributed evenly between change outputs.
+    | Proportional
+    -- ^ Value is distributed between change outputs in proportion to payment
+    -- outputs. Use this with v'RandomImprove' for /emergent/ management of the
+    -- wallet's UTxO.
+    | Unbalanced
+    -- ^ Value is added to the first change output until it reaches the maximum
+    -- size. Then the next change output will receive value.
+    deriving (Show, Read, Eq, Enum, Generic)
+
+-- | Placeholder for the type of possible failure cases in coin selection.
+newtype SelectionError = SelectionError String -- ^ Something went wrong.
+    deriving (Show, Read, Eq, Generic)
+
+-- $naming
+-- We call the coin selection function 'xyzzy' in lieu of a better name.
+--
+-- Possible names could be @balanceTransaction@, @selectCoins@, or
+-- something like that.
+
+-- | The top-level coin selection function.
+--
+-- Given an initial partially funded transaction body, it will produce a
+-- complete, funded, and balanced transaction body from the provided 'UTxO'.
+--
+-- If successful, it returns one or more transactions, otherwise the reason for
+-- failure is provided.
+--
+-- __How to use__
+--
+-- The main interface of 'xyzzy' is its parameter @tx@ the partial
+-- transaction. The caller specifies the outputs of @tx@, and 'xyzzy' calculates
+-- a balanced 'TxBody' from that.
+--
+-- [Payment]: Add a 'TxOut' to the initial @tx@ with the recipient
+--   'Address' and 'Value'.
+-- [Change Output]: Add a 'TxOut' to the initial @tx@ such that the 'Address'
+--   satisfies 'isOurAddress'. Its 'Value' would usually be zero.
+--   If the maximum output size is reached, another 'TxOut' with the same
+--   address will be added to the transaction.
+-- [No new change addresses]: If there is no transaction output that satisfies
+--   v'IsOurAddress', then change will be directed back to the addresses
+--   belonging to the wallet which funded the transaction.
```

Comment:
I like this idea of specifying a skeleton for transaction outputs, and filling it up. However, I feel that the simplicity of the scheme is lost a bit if we allow the algorithm to add new outputs — perhaps the function should even return an error in this case? 🤔 That said, I like Sam's observation that the funds can always be funneled through a total `Amount`, and only then be redistributed to transaction outputs. Then again, this "total amount" funnel does not consider fees, which *do* depend on the number of transaction outputs? 🤔